### DISCUSSION TITLE
Best approach to extend fabric classes

### DISCUSSION THREAD

**Discussion by bitttttten** [fabricjs/fabric.js#9777] [General]
Hey! Love the project and I just have a few questions. I'm looking to extend the base fabric object to always add an ID.

Later on I adjust the controls, and seems to break it.

```ts
class FabricObject extends fabric.Object {
  public id?: string;
  constructor(options: fabric.IObjectOptions) {
    super(options);
    this.id = `Math.random()`;
  }
}

fabric.Object = FabricObject;

// later on
fabric.Object.prototype.transparentCorners = false;
fabric.Object.prototype.cornerColor = 'blue';
fabric.Object.prototype.cornerStyle = 'circle';
```

I would imagine I can do something like this, although it doesn't seem to work exactly.

For example if I now try and set transparentCorners or cornerColor, this no longer works: https://stackblitz.com/edit/vitejs-vite-ahqkv6?file=src%2Fmain.ts.

What's the right approach here?

**Comment by asturur** [comment]
With which version of fabricJS are you working?

**Reply by bitttttten** [reply]
I'm on 5.3.0 @asturur

**Reply by asturur** [reply]
so extending the base class is a bit convolute with both versions.
What i would do is something like:

```
const originalInitialize = fabric.Object.prototype.initialize;
// do not remember if the initialize signature has just options as a single argument, please check.
fabric.Object.prototype.initialize = function (options) {
  originalInitialize.call(this, options);
  this.id = Math.random();
} 
fabric.Object.prototype.transparentCorners = false;
fabric.Object.prototype.cornerColor = 'blue';
fabric.Object.prototype.cornerStyle = 'circle';

```

**Reply by asturur** [reply]
in your example when you do:
```
fabric.Object = FabricObject;
```
all the other classes like Rect, Circle, etc etc are already created when you run your code, so swapping fabricObject in that moment isn't really having any effect

**Reply by bitttttten** [reply]
gotcha okay! that was my instinct so thanks for validating that. i will go ahead with your suggestion. will this change in fabric v6?

**Reply by asturur** [reply]
Yes a bit.
But not extremely. You will still be able to do it.

Is a long explanation of the difference between classes and functions with prototype and what TS does.
We will have to write it in the changelog and migration guide for fabric so i won't duplicate the effort now

**Reply by bitttttten** [reply]
yep sounds fair, was more gauging the effort as in should i wait or go with this approach, as there are many things i want to adjust and tweak 🧙

i will just go for it and refactor if necessary 

appreciate the replies! fabric6 is looking v nice

**Reply by asturur** [reply]
You should consider fabric 6 stable and just go with it.

**Reply by bitttttten** [reply]
how would you do this in fabric6?

if i do

```
import { FabricObject } from 'fabric'

FabricObject.prototype.controls.ml = ml;
```

i end up getting 

> Uncaught TypeError: Cannot set properties of undefined (setting 'ml') at index.ts:135:33

still trying to work out what i am doing wrong, but thought i'd ask as i hit a dead end with debugging :)

**Reply by asturur** [reply]
Controls are not on the prototype anymore in fabric6, at least not by default.

If you need to change the controls for all objects, wait a second for this PR to be merged:
https://github.com/fabricjs/fabric.js/pull/9891
In which i had to make some last minute change because there were some issues with the performances.

There is a function 'createControls' you can change that function to return the controls for all the objects.

**Reply by bitttttten** [reply]
we are on 6.0.0-rc4 and i'm still not really able to get controls working.
my code looks pretty much exactly like this: https://gist.github.com/bitttttten/26163d05bbfec4a7ac773322369aba4b
is that right?

```js
import { FabricObject } from 'fabric'

FabricObject.createControls = function () {
  return {
    controls: {
      ml,
      mr,
      mt,
      mb,
      tl,
      tr,
      bl,
      br,
      mtr,
    },
  }
}
```

**Reply by bitttttten** [reply]
actually, this works 🥳

**Comment by bitttttten** [comment]
Solved by doing something like:

```
const _initialize = fabric.Object.prototype.initialize

fabric.Object.prototype.initialize = function (...args) {
  _initialize.apply(this, args)
  this.someThingAppliedToAllObjects = true

  return this
}
```
